File: src/python/allocation.py
"""
Resource Allocation Schema - Best Practices Implementation

Implements allocation strategies for CPU, GPU, memory, and budget distribution
following established patterns: pool allocation, slab allocation, buddy system
"""

import time
import threading
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, field
from enum import Enum
from collections import OrderedDict
import heapq


class AllocationStrategy(Enum):
    """Allocation strategy types."""
    FIRST_FIT = "first_fit"      # Allocate first available block
    BEST_FIT = "best_fit"        # Allocate smallest sufficient block
    WORST_FIT = "worst_fit"      # Allocate largest available block
    POOL = "pool"                # Fixed-size block allocation
    SLAB = "slab"                # Object-cached allocation
    BUDDY = "buddy"              # Power-of-two block splitting


class ResourceType(Enum):
    """Resource type being allocated."""
    CPU_CORES = "cpu_cores"
    CPU_TIME = "cpu_time"
    GPU_COMPUTE = "gpu_compute"
    GPU_MEMORY = "gpu_memory"
    SYSTEM_MEMORY = "system_memory"
    BANDWIDTH = "bandwidth"
    POWER_BUDGET = "power_budget"
    THERMAL_HEADROOM = "thermal_headroom"


class Priority(Enum):
    """Allocation priority levels."""
    CRITICAL = 0    # System-critical, cannot be preempted
    HIGH = 1        # High priority, preempts normal
    NORMAL = 2      # Default priority
    LOW = 3         # Background tasks
    BEST_EFFORT = 4 # Allocate if available


@dataclass
class AllocationConstraints:
    """Constraints for allocation."""
    min_amount: Optional[int] = None
    max_amount: Optional[int] = None
    alignment: Optional[int] = None
    affinity: Optional[List[int]] = None        # Preferred cores/nodes
    anti_affinity: Optional[List[str]] = None   # Don't colocate with these
    timeout_ms: Optional[int] = None
    preemptible: bool = True


@dataclass
class AllocationRequest:
    """Allocation request."""
    id: str
    resource_type: ResourceType
    amount: int
    priority: Priority = Priority.NORMAL
    constraints: AllocationConstraints = field(default_factory=AllocationConstraints)
    metadata: Dict[str, str] = field(default_factory=dict)


@dataclass
class Allocation:
    """Allocation result."""
    id: str
    request_id: str
    resource_type: ResourceType
    amount: int
    block_id: int
    offset: int
    timestamp: float
    expires_at: Optional[float] = None


@dataclass
class Block:
    """Block in the allocation pool."""
    id: int
    offset: int
    size: int
    allocated: bool = False
    allocation_id: Optional[str] = None
    priority: Priority = Priority.BEST_EFFORT


@dataclass
class PoolStats:
    """Pool statistics."""
    total_capacity: int
    allocated: int
    available: int
    utilization: float
    fragmentation: float
    block_count: int
    free_block_count: int
    allocation_count: int
    largest_free_block: int


class AllocationError(Exception):
    """Allocation error base class."""
    pass


class InsufficientResourcesError(AllocationError):
    pass


class NotFoundError(AllocationError):
    pass


class InvalidConstraintsError(AllocationError):
    pass


class ResourcePool:
    """Resource pool for a single resource type."""

    def __init__(
        self,
        resource_type: ResourceType,
        capacity: int,
        strategy: AllocationStrategy = AllocationStrategy.BEST_FIT,
    ):
        self.resource_type = resource_type
        self.total_capacity = capacity
        self.strategy = strategy
        self.blocks: OrderedDict[int, Block] = OrderedDict()  # offset -> block
        self.allocations: Dict[str, Allocation] = {}
        self._next_block_id = 1
        self._lock = threading.RLock()

        # Initialize with single free block
        self.blocks[0] = Block(id=0, offset=0, size=capacity)

    def allocate(self, request: AllocationRequest) -> Allocation:
        """Allocate resources."""
        with self._lock:
            amount = self._apply_constraints(request.amount, request.constraints)

            # Find suitable block based on strategy
            block_offset = self._find_block(amount)
            if block_offset is None:
                raise InsufficientResourcesError(
                    f"Cannot allocate {amount} of {self.resource_type.value}"
                )

            block = self.blocks[block_offset]

            # Split block if necessary
            if block.size > amount:
                self._split_block(block_offset, amount)

            # Update block
            block = self.blocks[block_offset]
            allocation_id = f"alloc-{request.id}-{self._next_block_id}"
            block.allocated = True
            block.allocation_id = allocation_id
            block.priority = request.priority

            # Create allocation record
            now = time.time()
            expires_at = None
            if request.constraints.timeout_ms:
                expires_at = now + request.constraints.timeout_ms / 1000.0

            allocation = Allocation(
                id=allocation_id,
                request_id=request.id,
                resource_type=self.resource_type,
                amount=amount,
                block_id=block.id,
                offset=block_offset,
                timestamp=now,
                expires_at=expires_at,
            )

            self.allocations[allocation_id] = allocation
            self._next_block_id += 1

            return allocation

    def free(self, allocation_id: str) -> None:
        """Free an allocation."""
        with self._lock:
            allocation = self.allocations.pop(allocation_id, None)
            if not allocation:
                raise NotFoundError(f"Allocation {allocation_id} not found")

            block = self.blocks.get(allocation.offset)
            if block:
                block.allocated = False
                block.allocation_id = None
                self._coalesce_blocks(allocation.offset)

    def stats(self) -> PoolStats:
        """Get pool statistics."""
        with self._lock:
            allocated = sum(b.size for b in self.blocks.values() if b.allocated)
            free_blocks = [b for b in self.blocks.values() if not b.allocated]
            largest_free = max((b.size for b in free_blocks), default=0)

            return PoolStats(
                total_capacity=self.total_capacity,
                allocated=allocated,
                available=self.total_capacity - allocated,
                utilization=allocated / self.total_capacity if self.total_capacity > 0 else 0,
                fragmentation=self._calculate_fragmentation(),
                block_count=len(self.blocks),
                free_block_count=len(free_blocks),
                allocation_count=len(self.allocations),
                largest_free_block=largest_free,
            )

    def _apply_constraints(self, amount: int, constraints: AllocationConstraints) -> int:
        """Apply constraints to requested amount."""
        result = amount

        if constraints.min_amount:
            result = max(result, constraints.min_amount)
        if constraints.max_amount:
            result = min(result, constraints.max_amount)
        if constraints.alignment:
            result = ((result + constraints.alignment - 1) // constraints.alignment) * constraints.alignment

        return result

    def _find_block(self, size: int) -> Optional[int]:
        """Find a suitable block based on strategy."""
        free_blocks = [(o, b) for o, b in self.blocks.items() if not b.allocated and b.size >= size]

        if not free_blocks:
            return None

        if self.strategy == AllocationStrategy.FIRST_FIT:
            return free_blocks[0][0]
        elif self.strategy == AllocationStrategy.BEST_FIT:
            return min(free_blocks, key=lambda x: x[1].size)[0]
        elif self.strategy == AllocationStrategy.WORST_FIT:
            return max(free_blocks, key=lambda x: x[1].size)[0]
        elif self.strategy == AllocationStrategy.BUDDY:
            # Round up to power of two
            size = 1 << (size - 1).bit_length()
            suitable = [(o, b) for o, b in free_blocks if b.size >= size]
            return min(suitable, key=lambda x: x[1].size)[0] if suitable else None
        else:
            return free_blocks[0][0]

    def _split_block(self, offset: int, size: int) -> None:
        """Split a block into allocated and free portions."""
        block = self.blocks[offset]
        remaining = block.size - size

        # Resize original block
        block.size = size

        # Create new free block
        new_offset = offset + size
        self.blocks[new_offset] = Block(
            id=self._next_block_id,
            offset=new_offset,
            size=remaining,
        )
        self._next_block_id += 1

        # Re-sort blocks by offset
        self.blocks = OrderedDict(sorted(self.blocks.items()))

    def _coalesce_blocks(self, offset: int) -> None:
        """Coalesce adjacent free blocks."""
        offsets = list(self.blocks.keys())
        idx = offsets.index(offset)

        # Try to merge with next block
        if idx + 1 < len(offsets):
            next_offset = offsets[idx + 1]
            next_block = self.blocks[next_offset]
            current_block = self.blocks[offset]

            if not next_block.allocated and current_block.offset + current_block.size == next_offset:
                current_block.size += next_block.size
                del self.blocks[next_offset]
                offsets = list(self.blocks.keys())
                idx = offsets.index(offset)

        # Try to merge with previous block
        if idx > 0:
            prev_offset = offsets[idx - 1]
            prev_block = self.blocks[prev_offset]
            current_block = self.blocks[offset]

            if not prev_block.allocated and prev_block.offset + prev_block.size == offset:
                prev_block.size += current_block.size
                del self.blocks[offset]

    def _calculate_fragmentation(self) -> float:
        """Calculate memory fragmentation."""
        free_blocks = [b for b in self.blocks.values() if not b.allocated]
        if not free_blocks:
            return 0.0

        total_free = sum(b.size for b in free_blocks)
        largest_free = max(b.size for b in free_blocks)

        if total_free == 0:
            return 0.0

        return 1.0 - (largest_free / total_free)


class Allocator:
    """Multi-resource allocator."""

    def __init__(self):
        self.pools: Dict[ResourceType, ResourcePool] = {}
        self.allocation_log: List[Dict[str, Any]] = []
        self._lock = threading.RLock()

    def init_pool(
        self,
        resource_type: ResourceType,
        capacity: int,
        strategy: AllocationStrategy = AllocationStrategy.BEST_FIT,
    ) -> None:
        """Initialize a resource pool."""
        with self._lock:
            self.pools[resource_type] = ResourcePool(resource_type, capacity, strategy)

    def allocate(self, request: AllocationRequest) -> Allocation:
        """Allocate from a specific pool."""
        with self._lock:
            pool = self.pools.get(request.resource_type)
            if not pool:
                raise InvalidConstraintsError(f"No pool for {request.resource_type.value}")

            allocation = pool.allocate(request)

            self.allocation_log.append({
                "timestamp": time.time(),
                "event": "allocated",
                "allocation_id": allocation.id,
                "resource_type": request.resource_type.value,
                "amount": allocation.amount,
            })

            return allocation

    def free(self, resource_type: ResourceType, allocation_id: str) -> None:
        """Free an allocation."""
        with self._lock:
            pool = self.pools.get(resource_type)
            if not pool:
                raise NotFoundError(f"No pool for {resource_type.value}")

            pool.free(allocation_id)

            self.allocation_log.append({
                "timestamp": time.time(),
                "event": "freed",
                "allocation_id": allocation_id,
                "resource_type": resource_type.value,
            })

    def stats(self) -> Dict[ResourceType, PoolStats]:
        """Get stats for all pools."""
        with self._lock:
            return {rt: pool.stats() for rt, pool in self.pools.items()}

    def cleanup_expired(self) -> int:
        """Cleanup expired allocations. Returns count of cleaned allocations."""
        count = 0
        now = time.time()

        with self._lock:
            for resource_type, pool in self.pools.items():
                expired = [
                    alloc_id
                    for alloc_id, alloc in pool.allocations.items()
                    if alloc.expires_at and alloc.expires_at < now
                ]

                for alloc_id in expired:
                    try:
                        pool.free(alloc_id)
                        self.allocation_log.append({
                            "timestamp": now,
                            "event": "expired",
                            "allocation_id": alloc_id,
                            "resource_type": resource_type.value,
                        })
                        count += 1
                    except AllocationError:
                        pass

        return count

    def get_log(self) -> List[Dict[str, Any]]:
        """Get allocation log."""
        return self.allocation_log.copy()


# Convenience functions
def create_default_allocator() -> Allocator:
    """Create an allocator with default pools."""
    allocator = Allocator()
    allocator.init_pool(ResourceType.CPU_CORES, 16, AllocationStrategy.FIRST_FIT)
    allocator.init_pool(ResourceType.CPU_TIME, 100000, AllocationStrategy.BEST_FIT)
    allocator.init_pool(ResourceType.GPU_MEMORY, 8192, AllocationStrategy.BUDDY)
    allocator.init_pool(ResourceType.SYSTEM_MEMORY, 32768, AllocationStrategy.BEST_FIT)
    allocator.init_pool(ResourceType.POWER_BUDGET, 100, AllocationStrategy.FIRST_FIT)
    return allocator


if __name__ == "__main__":
    # Demo
    allocator = create_default_allocator()

    # Allocate CPU cores
    cpu_req = AllocationRequest(
        id="task-1",
        resource_type=ResourceType.CPU_CORES,
        amount=4,
        priority=Priority.HIGH,
    )
    cpu_alloc = allocator.allocate(cpu_req)
    print(f"Allocated CPU: {cpu_alloc}")

    # Allocate memory
    mem_req = AllocationRequest(
        id="task-1",
        resource_type=ResourceType.SYSTEM_MEMORY,
        amount=4096,
        priority=Priority.NORMAL,
        constraints=AllocationConstraints(alignment=64),
    )
    mem_alloc = allocator.allocate(mem_req)
    print(f"Allocated Memory: {mem_alloc}")

    # Print stats
    print("\nPool Statistics:")
    for rt, stats in allocator.stats().items():
        print(f"  {rt.value}: {stats.allocated}/{stats.total_capacity} "
              f"({stats.utilization*100:.1f}% util, {stats.fragmentation*100:.1f}% frag)")

    # Free allocations
    allocator.free(ResourceType.CPU_CORES, cpu_alloc.id)
    allocator.free(ResourceType.SYSTEM_MEMORY, mem_alloc.id)

    print("\nAfter freeing:")
    for rt, stats in allocator.stats().items():
        print(f"  {rt.value}: {stats.allocated}/{stats.total_capacity}")

File: src/rust/allocation.rs
//! Resource Allocation Schema - Best Practices Implementation
//!
//! Implements allocation strategies for CPU, GPU, memory, and budget distribution
//! following established patterns: pool allocation, slab allocation, buddy system

use std::collections::{BTreeMap, HashMap, VecDeque};
use std::sync::{Arc, RwLock};
use serde::{Deserialize, Serialize};

/// Allocation strategy types
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum AllocationStrategy {
    /// First-fit: allocate first available block
    FirstFit,
    /// Best-fit: allocate smallest sufficient block
    BestFit,
    /// Worst-fit: allocate largest available block
    WorstFit,
    /// Pool: fixed-size block allocation
    Pool,
    /// Slab: object-cached allocation
    Slab,
    /// Buddy: power-of-two block splitting
    Buddy,
}

/// Resource type being allocated
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum ResourceType {
    CpuCores,
    CpuTime,
    GpuCompute,
    GpuMemory,
    SystemMemory,
    Bandwidth,
    PowerBudget,
    ThermalHeadroom,
}

/// Allocation request
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AllocationRequest {
    pub id: String,
    pub resource_type: ResourceType,
    pub amount: u64,
    pub priority: Priority,
    pub constraints: AllocationConstraints,
    pub metadata: HashMap<String, String>,
}

/// Allocation priority levels
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub enum Priority {
    Critical = 0,    // System-critical, cannot be preempted
    High = 1,        // High priority, preempts normal
    Normal = 2,      // Default priority
    Low = 3,         // Background tasks
    BestEffort = 4,  // Allocate if available
}

/// Constraints for allocation
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct AllocationConstraints {
    pub min_amount: Option<u64>,
    pub max_amount: Option<u64>,
    pub alignment: Option<u64>,
    pub affinity: Option<Vec<u32>>,  // Preferred cores/nodes
    pub anti_affinity: Option<Vec<String>>,  // Don't colocate with these
    pub timeout_ms: Option<u64>,
    pub preemptible: bool,
}

/// Allocation result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Allocation {
    pub id: String,
    pub request_id: String,
    pub resource_type: ResourceType,
    pub amount: u64,
    pub block_id: u64,
    pub offset: u64,
    pub timestamp: u64,
    pub expires_at: Option<u64>,
}

/// Block in the allocation pool
#[derive(Debug, Clone)]
struct Block {
    id: u64,
    offset: u64,
    size: u64,
    allocated: bool,
    allocation_id: Option<String>,
    priority: Priority,
}

/// Resource pool for a single resource type
pub struct ResourcePool {
    resource_type: ResourceType,
    total_capacity: u64,
    strategy: AllocationStrategy,
    blocks: BTreeMap<u64, Block>,  // offset -> block
    free_list: VecDeque<u64>,      // free block offsets
    allocations: HashMap<String, Allocation>,
    next_block_id: u64,
    fragmentation_threshold: f64,
}

impl ResourcePool {
    pub fn new(resource_type: ResourceType, capacity: u64, strategy: AllocationStrategy) -> Self {
        let mut pool = Self {
            resource_type,
            total_capacity: capacity,
            strategy,
            blocks: BTreeMap::new(),
            free_list: VecDeque::new(),
            allocations: HashMap::new(),
            next_block_id: 1,
            fragmentation_threshold: 0.3,
        };

        // Initialize with single free block
        pool.blocks.insert(0, Block {
            id: 0,
            offset: 0,
            size: capacity,
            allocated: false,
            allocation_id: None,
            priority: Priority::BestEffort,
        });
        pool.free_list.push_back(0);

        pool
    }

    /// Allocate resources
    pub fn allocate(&mut self, request: &AllocationRequest) -> Result<Allocation, AllocationError> {
        let amount = self.apply_constraints(request.amount, &request.constraints);

        let block_offset = match self.strategy {
            AllocationStrategy::FirstFit => self.first_fit(amount),
            AllocationStrategy::BestFit => self.best_fit(amount),
            AllocationStrategy::WorstFit => self.worst_fit(amount),
            AllocationStrategy::Pool => self.pool_allocate(amount),
            AllocationStrategy::Buddy => self.buddy_allocate(amount),
            AllocationStrategy::Slab => self.slab_allocate(amount),
        }?;

        let block = self.blocks.get_mut(&block_offset).unwrap();
        let allocation_id = format!("alloc-{}-{}", request.id, self.next_block_id);

        // Split block if necessary
        if block.size > amount {
            self.split_block(block_offset, amount);
        }

        let block = self.blocks.get_mut(&block_offset).unwrap();
        block.allocated = true;
        block.allocation_id = Some(allocation_id.clone());
        block.priority = request.priority;

        let allocation = Allocation {
            id: allocation_id.clone(),
            request_id: request.id.clone(),
            resource_type: self.resource_type,
            amount,
            block_id: block.id,
            offset: block_offset,
            timestamp: current_timestamp(),
            expires_at: request.constraints.timeout_ms.map(|t| current_timestamp() + t),
        };

        self.allocations.insert(allocation_id, allocation.clone());
        self.next_block_id += 1;

        Ok(allocation)
    }

    /// Free an allocation
    pub fn free(&mut self, allocation_id: &str) -> Result<(), AllocationError> {
        let allocation = self.allocations.remove(allocation_id)
            .ok_or(AllocationError::NotFound)?;

        let block = self.blocks.get_mut(&allocation.offset)
            .ok_or(AllocationError::InvalidBlock)?;

        block.allocated = false;
        block.allocation_id = None;

        self.free_list.push_back(allocation.offset);
        self.coalesce_blocks(allocation.offset);

        Ok(())
    }

    /// Get pool statistics
    pub fn stats(&self) -> PoolStats {
        let allocated: u64 = self.blocks.values()
            .filter(|b| b.allocated)
            .map(|b| b.size)
            .sum();

        let free_blocks: usize = self.blocks.values()
            .filter(|b| !b.allocated)
            .count();

        let largest_free = self.blocks.values()
            .filter(|b| !b.allocated)
            .map(|b| b.size)
            .max()
            .unwrap_or(0);

        PoolStats {
            total_capacity: self.total_capacity,
            allocated,
            available: self.total_capacity - allocated,
            utilization: allocated as f64 / self.total_capacity as f64,
            fragmentation: self.calculate_fragmentation(),
            block_count: self.blocks.len(),
            free_block_count: free_blocks,
            allocation_count: self.allocations.len(),
            largest_free_block: largest_free,
        }
    }

    fn apply_constraints(&self, amount: u64, constraints: &AllocationConstraints) -> u64 {
        let mut result = amount;

        if let Some(min) = constraints.min_amount {
            result = result.max(min);
        }
        if let Some(max) = constraints.max_amount {
            result = result.min(max);
        }
        if let Some(align) = constraints.alignment {
            result = (result + align - 1) / align * align;
        }

        result
    }

    fn first_fit(&self, size: u64) -> Result<u64, AllocationError> {
        for (&offset, block) in &self.blocks {
            if !block.allocated && block.size >= size {
                return Ok(offset);
            }
        }
        Err(AllocationError::InsufficientResources)
    }

    fn best_fit(&self, size: u64) -> Result<u64, AllocationError> {
        self.blocks.iter()
            .filter(|(_, b)| !b.allocated && b.size >= size)
            .min_by_key(|(_, b)| b.size)
            .map(|(&offset, _)| offset)
            .ok_or(AllocationError::InsufficientResources)
    }

    fn worst_fit(&self, size: u64) -> Result<u64, AllocationError> {
        self.blocks.iter()
            .filter(|(_, b)| !b.allocated && b.size >= size)
            .max_by_key(|(_, b)| b.size)
            .map(|(&offset, _)| offset)
            .ok_or(AllocationError::InsufficientResources)
    }

    fn pool_allocate(&self, size: u64) -> Result<u64, AllocationError> {
        // Pool allocation: fixed-size blocks, find any free
        self.first_fit(size)
    }

    fn buddy_allocate(&mut self, size: u64) -> Result<u64, AllocationError> {
        // Buddy system: round up to power of two
        let size = size.next_power_of_two();
        self.best_fit(size)
    }

    fn slab_allocate(&self, size: u64) -> Result<u64, AllocationError> {
        // Slab: use best-fit for object caching
        self.best_fit(size)
    }

    fn split_block(&mut self, offset: u64, size: u64) {
        let block = self.blocks.get(&offset).unwrap().clone();
        let remaining = block.size - size;

        // Update original block
        self.blocks.get_mut(&offset).unwrap().size = size;

        // Create new free block
        let new_offset = offset + size;
        self.blocks.insert(new_offset, Block {
            id: self.next_block_id,
            offset: new_offset,
            size: remaining,
            allocated: false,
            allocation_id: None,
            priority: Priority::BestEffort,
        });
        self.next_block_id += 1;
        self.free_list.push_back(new_offset);
    }

    fn coalesce_blocks(&mut self, offset: u64) {
        // Coalesce with next block
        let block = self.blocks.get(&offset).unwrap().clone();
        let next_offset = offset + block.size;

        if let Some(next_block) = self.blocks.get(&next_offset) {
            if !next_block.allocated {
                let next_size = next_block.size;
                self.blocks.remove(&next_offset);
                self.blocks.get_mut(&offset).unwrap().size += next_size;
                self.free_list.retain(|&o| o != next_offset);
            }
        }

        // Coalesce with previous block
        if let Some((&prev_offset, prev_block)) = self.blocks.range(..offset).last() {
            if !prev_block.allocated && prev_offset + prev_block.size == offset {
                let current_size = self.blocks.get(&offset).unwrap().size;
                self.blocks.remove(&offset);
                self.blocks.get_mut(&prev_offset).unwrap().size += current_size;
                self.free_list.retain(|&o| o != offset);
            }
        }
    }

    fn calculate_fragmentation(&self) -> f64 {
        let free_blocks: Vec<_> = self.blocks.values()
            .filter(|b| !b.allocated)
            .collect();

        if free_blocks.is_empty() {
            return 0.0;
        }

        let total_free: u64 = free_blocks.iter().map(|b| b.size).sum();
        let largest_free = free_blocks.iter().map(|b| b.size).max().unwrap_or(0);

        if total_free == 0 {
            return 0.0;
        }

        1.0 - (largest_free as f64 / total_free as f64)
    }
}

/// Pool statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PoolStats {
    pub total_capacity: u64,
    pub allocated: u64,
    pub available: u64,
    pub utilization: f64,
    pub fragmentation: f64,
    pub block_count: usize,
    pub free_block_count: usize,
    pub allocation_count: usize,
    pub largest_free_block: u64,
}

/// Allocation errors
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AllocationError {
    InsufficientResources,
    InvalidConstraints,
    NotFound,
    InvalidBlock,
    Timeout,
    Preempted,
}

/// Multi-resource allocator
pub struct Allocator {
    pools: HashMap<ResourceType, Arc<RwLock<ResourcePool>>>,
    allocation_log: Vec<AllocationEvent>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AllocationEvent {
    pub timestamp: u64,
    pub event_type: AllocationEventType,
    pub allocation_id: String,
    pub resource_type: ResourceType,
    pub amount: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum AllocationEventType {
    Allocated,
    Freed,
    Preempted,
    Expired,
    Resized,
}

impl Allocator {
    pub fn new() -> Self {
        Self {
            pools: HashMap::new(),
            allocation_log: Vec::new(),
        }
    }

    /// Initialize a resource pool
    pub fn init_pool(
        &mut self,
        resource_type: ResourceType,
        capacity: u64,
        strategy: AllocationStrategy,
    ) {
        let pool = ResourcePool::new(resource_type, capacity, strategy);
        self.pools.insert(resource_type, Arc::new(RwLock::new(pool)));
    }

    /// Allocate from a specific pool
    pub fn allocate(&mut self, request: AllocationRequest) -> Result<Allocation, AllocationError> {
        let pool = self.pools.get(&request.resource_type)
            .ok_or(AllocationError::InvalidConstraints)?;

        let allocation = pool.write().unwrap().allocate(&request)?;

        self.allocation_log.push(AllocationEvent {
            timestamp: current_timestamp(),
            event_type: AllocationEventType::Allocated,
            allocation_id: allocation.id.clone(),
            resource_type: request.resource_type,
            amount: allocation.amount,
        });

        Ok(allocation)
    }

    /// Free an allocation
    pub fn free(&mut self, resource_type: ResourceType, allocation_id: &str) -> Result<(), AllocationError> {
        let pool = self.pools.get(&resource_type)
            .ok_or(AllocationError::NotFound)?;

        pool.write().unwrap().free(allocation_id)?;

        self.allocation_log.push(AllocationEvent {
            timestamp: current_timestamp(),
            event_type: AllocationEventType::Freed,
            allocation_id: allocation_id.to_string(),
            resource_type,
            amount: 0,
        });

        Ok(())
    }

    /// Get stats for all pools
    pub fn stats(&self) -> HashMap<ResourceType, PoolStats> {
        self.pools.iter()
            .map(|(rt, pool)| (*rt, pool.read().unwrap().stats()))
            .collect()
    }

    /// Get allocation log
    pub fn get_log(&self) -> &[AllocationEvent] {
        &self.allocation_log
    }

    /// Cleanup expired allocations
    pub fn cleanup_expired(&mut self) {
        let now = current_timestamp();

        for (resource_type, pool) in &self.pools {
            let mut pool = pool.write().unwrap();
            let expired: Vec<_> = pool.allocations.iter()
                .filter(|(_, a)| a.expires_at.map(|e| e < now).unwrap_or(false))
                .map(|(id, _)| id.clone())
                .collect();

            for id in expired {
                if pool.free(&id).is_ok() {
                    self.allocation_log.push(AllocationEvent {
                        timestamp: now,
                        event_type: AllocationEventType::Expired,
                        allocation_id: id,
                        resource_type: *resource_type,
                        amount: 0,
                    });
                }
            }
        }
    }
}

impl Default for Allocator {
    fn default() -> Self {
        Self::new()
    }
}

fn current_timestamp() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis() as u64
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_allocation() {
        let mut pool = ResourcePool::new(ResourceType::SystemMemory, 1000, AllocationStrategy::FirstFit);

        let request = AllocationRequest {
            id: "test-1".into(),
            resource_type: ResourceType::SystemMemory,
            amount: 100,
            priority: Priority::Normal,
            constraints: AllocationConstraints::default(),
            metadata: HashMap::new(),
        };

        let allocation = pool.allocate(&request).unwrap();
        assert_eq!(allocation.amount, 100);

        let stats = pool.stats();
        assert_eq!(stats.allocated, 100);
        assert_eq!(stats.available, 900);
    }

    #[test]
    fn test_allocation_and_free() {
        let mut pool = ResourcePool::new(ResourceType::CpuTime, 1000, AllocationStrategy::BestFit);

        let req1 = AllocationRequest {
            id: "req-1".into(),
            resource_type: ResourceType::CpuTime,
            amount: 200,
            priority: Priority::High,
            constraints: AllocationConstraints::default(),
            metadata: HashMap::new(),
        };

        let alloc1 = pool.allocate(&req1).unwrap();
        assert_eq!(pool.stats().allocated, 200);

        pool.free(&alloc1.id).unwrap();
        assert_eq!(pool.stats().allocated, 0);
    }

    #[test]
    fn test_buddy_allocation() {
        let mut pool = ResourcePool::new(ResourceType::GpuMemory, 1024, AllocationStrategy::Buddy);

        let request = AllocationRequest {
            id: "gpu-1".into(),
            resource_type: ResourceType::GpuMemory,
            amount: 100,  // Will be rounded to 128 (next power of 2)
            priority: Priority::Normal,
            constraints: AllocationConstraints::default(),
            metadata: HashMap::new(),
        };

        let allocation = pool.allocate(&request).unwrap();
        assert!(allocation.amount >= 100);
    }

    #[test]
    fn test_multi_resource_allocator() {
        let mut allocator = Allocator::new();
        allocator.init_pool(ResourceType::CpuCores, 16, AllocationStrategy::FirstFit);
        allocator.init_pool(ResourceType::SystemMemory, 16384, AllocationStrategy::BestFit);

        let cpu_req = AllocationRequest {
            id: "cpu-1".into(),
            resource_type: ResourceType::CpuCores,
            amount: 4,
            priority: Priority::Normal,
            constraints: AllocationConstraints::default(),
            metadata: HashMap::new(),
        };

        let mem_req = AllocationRequest {
            id: "mem-1".into(),
            resource_type: ResourceType::SystemMemory,
            amount: 4096,
            priority: Priority::Normal,
            constraints: AllocationConstraints::default(),
            metadata: HashMap::new(),
        };

        allocator.allocate(cpu_req).unwrap();
        allocator.allocate(mem_req).unwrap();

        let stats = allocator.stats();
        assert_eq!(stats[&ResourceType::CpuCores].allocated, 4);
        assert_eq!(stats[&ResourceType::SystemMemory].allocated, 4096);
    }
}
